api4/websocket_test.go

```diff
@@ -78,6 +78,122 @@ func TestWebSocket(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	WebSocketClient, err := th.CreateWebSocketClient()
+	require.Nil(t, err)
+	defer WebSocketClient.Close()
+
+	WebSocketClient.Listen()
+
+	for name, tc := range map[string]struct {
+		OldStatus            *model.Status
+		UserIsActive         bool
+		Manual               bool
+		ExpectedStatusString string
+	}{
+		"User set dnd manually": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_DND,
+				Manual: true,
+			},
+			UserIsActive:         true,
+			Manual:               true,
+			ExpectedStatusString: model.STATUS_DND,
+		},
+		"User set out of office manually": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_OUT_OF_OFFICE,
+				Manual: true,
+			},
+			UserIsActive:         true,
+			Manual:               true,
+			ExpectedStatusString: model.STATUS_OUT_OF_OFFICE,
+		},
+		"User set offline manually": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_OFFLINE,
+				Manual: true,
+			},
+			UserIsActive:         true,
+			Manual:               true,
+			ExpectedStatusString: model.STATUS_OFFLINE,
+		},
+		"App automatically set user offline, user is active": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_OFFLINE,
+				Manual: false,
+			},
+			UserIsActive:         true,
+			Manual:               false,
+			ExpectedStatusString: model.STATUS_ONLINE,
+		},
+		"App automatically set user offline, user is not active": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_OFFLINE,
+				Manual: false,
+			},
+			UserIsActive:         false,
+			Manual:               false,
+			ExpectedStatusString: model.STATUS_AWAY,
+		},
+		"User manually set away, user is active with manual true": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_AWAY,
+				Manual: true,
+			},
+			UserIsActive:         true,
+			Manual:               true,
+			ExpectedStatusString: model.STATUS_ONLINE,
+		},
+		"User manually set away, user is active with manual false": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_AWAY,
+				Manual: true,
+			},
+			UserIsActive:         true,
+			Manual:               false,
+			ExpectedStatusString: model.STATUS_AWAY,
+		},
+		"User automatically set away, user is active with manual false": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_AWAY,
+				Manual: false,
+			},
+			UserIsActive:         true,
+			Manual:               false,
+			ExpectedStatusString: model.STATUS_ONLINE,
+		},
+		"User was online, no longer active": {
+			OldStatus: &model.Status{
+				Status: model.STATUS_ONLINE,
+				Manual: false,
+			},
+			UserIsActive:         false,
+			Manual:               false,
+			ExpectedStatusString: model.STATUS_AWAY,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			userId := th.BasicUser.Id
+			tc.OldStatus.UserId = userId
+
+			th.App.SaveAndBroadcastStatus(tc.OldStatus)
+
+			WebSocketClient.UserUpdateActiveStatus(tc.UserIsActive, tc.Manual)
+			time.Sleep(300 * time.Millisecond)
```

Comment:
Do we actually need to wait at all? Wondering if the blocking read on `WebSocketClient.ResponseChannel` suffices.

If the timeout is necessary, this will almost certainly end up being flaky.

Reply:
Running the tests without the timeout causes the test to fail. Running the tests with a timeout of 10 ms seems to work well. Also, if I run the tests 8+ times in a loop, the test consistently lock up. Whereas running it 7 or less times in the loop, it passes just fine. Could this be a limitation on the number of concurrent connections? Maybe there needs to be some cleanup between each test.

Reply:
@lieut-data What is your stance on keeping the tests as they are currently, based on the findings above?

Reply:
@mickmister and I synced up offline, and I realized that this timeout was already a pattern used in the above tests. Let's just stay consistent here and if flakiness arises, it's likely going to affect the existing tests and we can address holistically at that point.